src/pratt2.rs

```diff
@@ -0,0 +1,296 @@
+                                    match self.pratt_go::<M, _, _, _>(inp, $X.associativity().right_power()) {
+                                        Ok(rhs) => break 'choice M::combine(op, rhs, |op, rhs| $X.fold_prefix(op, rhs)),
+                                        Err(()) => inp.rewind(pre_op),
+                                    }
+                                },
+                                Err(()) => inp.rewind(pre_op),
+                            }
+                        }
+                    )*
+
+                    self.atom.go::<M>(inp)?
+                };
+
+                'luup: loop {
+                    enum Op<$($X),*> { $($X($X)),* }
+
+                    let ($($X,)*) = &self.ops;
+
+                    let pre_op = inp.save();
+                    let op = 'choice: {
+                        $(
+                            if $X::INFIX {
+                                match $X.op_parser().go::<M>(inp) {
+                                    Ok(out) => break 'choice Op::$X(out),
+                                    Err(()) => inp.rewind(pre_op),
+                                }
+                            }
+                        )*
+
+                        inp.rewind(pre_op);
+                        break 'luup;
+                    };
+
+                    match op {
+                        $(
+                            Op::$X(op) => {
+                                let assoc = $X.associativity();
+                                if assoc.left_power() < min_power {
+                                    inp.rewind(pre_op);
+                                    break
+                                } else {
+                                    lhs = match self.pratt_go::<M, _, _, _>(inp, assoc.right_power()) {
+                                        Ok(rhs) => M::combine(
+                                            M::combine(lhs, rhs, |lhs, rhs| (lhs, rhs)),
+                                            op,
+                                            |(lhs, rhs), op| $X.fold_infix(lhs, op, rhs),
+                                        ),
+                                        Err(()) => { inp.rewind(pre_op); break },
+                                    }
+                                }
+                            },
+                        )*
+                    }
+                }
+
+                Ok(lhs)
+            }
+        }
+
+        #[allow(unused_variables, non_snake_case)]
+        impl<'a, I, O, E, Atom, $($X),*> ParserSealed<'a, I, O, E> for Pratt<Atom, ($($X,)*)>
+        where
+            I: Input<'a>,
+            E: ParserExtra<'a, I>,
+            Atom: Parser<'a, I, O, E>,
+            $($X: Operator<'a, I, O, E>),*
+        {
+            fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E>) -> PResult<M, O> {
+                self.pratt_go::<M, _, _, _>(inp, 0)
+            }
+
+            go_extra!(O);
+        }
+    };
+}
```

Comment:
Definitely a while, although it wasn't as a bad as you might imagine.

When writing these sorts of implement-for-many-tuples macros, I start off writing just a regular implementation for a case that's not too hard to write (usually 2-element tuples). Once I've implemented that and tested that it works, I lift it up into a macro and generalise it.

I also started off with only supporting infix operators. Took me a while to think of the associated constant trick, but everything sort of just... materialised from there on?

If you have ideas about how to make it easier to understands, I'd definitely appreciate your thoughts! I personally don't find it too bad, but I realise that there's always a vast gulf between the person that implements something (and hence understands the mindset and logic that led to its creation) vs someone trying to implement it from a standing start.

Reply:
I suppose it might warrant a tiny comment about the implementation? Something along the lines of
```
// Pratt parsing macro for arbitrary tuples.
// 
// Takes the tuple apart and first applies only the prefix parsers. If one matches, then we have the first part and can continue to the main loop. If not, then we parse an atom, and then continue to the main loop.
// ....
```